File: sample.m
classdef sample
    properties
        Y
        f
        centeridx
        Ycentered
        fcentered
        n
        m
        mgood
        distance
        % * Ycentered, fcentered, and distance do not include the center. 
    end
    
    methods
        
       function obj = sample(x, par)
           % initialization
           
            % input dimension
            obj.n = par.n;

            % initial sample
            Y = bsxfun(@plus, x', par.tr.delta * [eye(obj.n); -eye(obj.n)]);
%             Y = mvnrnd(x, par.tr.delta / 3 * eye(obj.n), par.sample.initial - 1);
            obj.Y = [x'; Y]; 

            % initial sample size
            obj.m = size(obj.Y,1);

            % number of sample points within the acceptable range
            obj.mgood = obj.m;

            %function value of initial sample
            obj.f = nan(obj.m, 1);

       end
       
       function obj = centering(obj, centeridx)
           % centering
           % This is also the second stage of the initialization.
           
            % centering origin
            obj.centeridx = centeridx;

            % centered sample
            obj.Ycentered = bsxfun(@minus, ...
                obj.Y([1:obj.centeridx-1, obj.centeridx+1:end], :), ...
                obj.Y(centeridx, :) );
            obj.fcentered = obj.f([1:obj.centeridx-1, obj.centeridx+1:end]) - obj.f(centeridx);

            % distance of each point in the sample (excluding the center)
            % to the center
            obj.distance = vecnorm(obj.Ycentered, 2, 2);
       end
       
       function [obj, indeces] = addpoints(obj, points, fvalues, delta, par)
           % add new points 
           % After adding the new point, the sample size may exceed
           % par.sample.max. 
           % Returns the new sample set and the indeces of the newly added
           % points in the sample set. 
           
            % number of new points
            p = size(points, 1);

            % sample size 
            obj.m = obj.m + p;

            % sample
            obj.Y = [obj.Y; points];

            % centered points 
            centeredPoints = bsxfun(@minus, points, obj.Y(obj.centeridx,:));
            obj.Ycentered = [obj.Ycentered; centeredPoints];

            % function values
            obj.f = [obj.f; fvalues];

            % centered function values
            obj.fcentered = [obj.fcentered; fvalues - obj.f(obj.centeridx)];

            % distance to the center
            d = vecnorm(centeredPoints, 2, 2);
            obj.distance = [obj.distance; d];

            % number of sample points within the acceptable range
            obj.mgood = obj.mgood + sum(d < delta * par.sample.toremove);
            
            % indeces of the new points
            indeces = (obj.m - p + 1 : obj.m)';
        end

       
       function [obj, indeces] = deletepoints(obj, delta, par)
            % remove points that are too far away from the center  
            % After the deletion, the sample size should be no less than
            % par.sample.min, and no more than par.sample_max.
            % Returns the new sample set and a list for referencing the new indeces. 
            % newindex = indeces(oldindex) 
            % This newindex is 0 is the the point corresponding to the
            % oldindex is deleted. 
            
            % indeces of the farthest ones; 
            % We should not delete too many points so that sample size goes under par.sample_min.
            [~, farthest] = maxk(obj.distance, obj.m - par.sample.min);
            
            % indeces of the ones that are too far
            toofar = find(obj.distance > par.sample.toremove * delta);
            
            % indeces of the points to be removed
            if isempty(toofar) && obj.m > par.sample.max
                % We need to keep the sample size under its maximum. 
                toremove = farthest(1:(obj.m - par.sample.max));
            else
                toremove = intersect(farthest, toofar);
            end
            
            % remove points
            obj.Ycentered(toremove, :) = [];
            obj.fcentered(toremove, :) = [];
            obj.distance(toremove, :) = [];
            toremove1 = toremove;
            toremove1(toremove >= obj.centeridx) = toremove1(toremove >= obj.centeridx) + 1;
            obj.Y(toremove1,:) = [];
            obj.f(toremove1,:) = [];

            % update sample size
            obj.m = size(obj.Y, 1);
            obj.mgood = obj.m + length(toremove) - length(toofar);
            
            % the indeces referencing list
            indeces = nan(obj.m + length(toremove), 1);
            removed = false(obj.m + length(toremove), 1);
            removed(toremove1) = true;
            indeces(removed) = 0;
            indeces(~removed) = 1:obj.m;
            
            % update the center index
            obj.centeridx = indeces(obj.centeridx);
       end

       function points = poisedness_improvement_points(obj, p, delta, par)
            % poisedness improvement points
            % p is the number of points requested 
            
            Y = [zeros(1, obj.n); obj.Ycentered];
            
            delta = 0.75 * delta * par.sample.toremove;
            
            if strcmp(par.algorithm.model, 'linear')
                % the features
                Phi = nan(obj.m, 1 + obj.n);
                % the Lagrangian polynomials
                A = eye(1 + obj.n, obj.m + p);
            elseif strcmp(par.algorithm.model, 'quadratic')
                % the features
                Phi = nan(obj.m, 1 + obj.n + (1+obj.n)*obj.n/2);
                % the Lagrangian polynomials
                A = eye(1 + obj.n + (1+obj.n)*obj.n/2, obj.m + p);
            end
            
            for i = 1:obj.m
                Phi(i,:) = phi(Y(i,:)', par.algorithm.model);
            end
            points = [];
            
            
            
            for i = 1:size(A, 2)
                lis = Phi * A(:,i);
    
                [v,ji] = max(abs(lis(i:end)));
                ji = ji + i - 1;
                if isempty(v), v = 0; end

                if v > 0.001 && i <= obj.m
                    Y([i, ji], :) = Y([ji, i], :);
                    Phi([i, ji], :) = Phi([ji, i], :);
 
                else
                    if i <= obj.m
                        a = 1;
                    end
                    if strcmp(par.algorithm.model, 'linear')
                        g = A(2:obj.n+1,i);
                        
                        y1 = - delta / norm(g) * g;
                        val1 = A(1,i) + g'*y1;
                        y2 = delta / norm(g) * g;
                        val2 = A(1,i) + g'*y2;
                        
                    elseif strcmp(par.algorithm.model, 'quadratic')
                        g = A(2:obj.n+1,i);
                        H = zeros(obj.n);
                        H(triu(true(obj.n))) = A(obj.n+2:end,i);
                        H = H + H';
                        [y1, val1] = trust(-g, -H, delta);
                        val1 = A(1,i) - val1;
                        [y2, val2] = trust(g, H, delta);
                        val2 = A(1,i) + val2;
                    end

                    if abs(val1) > abs(val2)
                        Y(i, :) = y1;
                        Phi(i, :) = phi(y1, par.algorithm.model);
                        points = [points; y1'];
                    else
                        Y(i, :) = y2;
                        Phi(i, :) = phi(y2, par.algorithm.model);
                        points = [points; y2'];
                    end
%                     break;
                end

                % normalization
                A(:,i) = A(:,i) / ( Phi(i,:) * A(:,i) );

                % orthogonalization
                A(:, i+1:size(A,2)) = A(:, i+1:size(A,2)) ...
                    - A(:,i) * (Phi(i,:) * A(:, i+1:size(A,2)));
            end
       end
    end
end

File: Python2_obsolete/sample.py
"""sample management"""
import numpy as np
from .build_model import quadraticModel
from .trust_sub import trust_sub

class sample:
    def __init__(self, x, func, par):
        # input dimension
        self.n = par.n

        # initial sample size
        self.m = par.sample_initial

        # number of sample points within the acceptable range
        self.mgood = self.m

        # initial sample
        np.random.seed(0)
        Y = np.random.rand(self.m - 1, par.n) * 2 - 1
        Y = par.tr_delta / np.sqrt(par.n) * Y
        self.Y = np.concatenate((x + Y, x.reshape(1,par.n) ))

        # function value of initial sample
        self.f = np.array( map(func, self.Y) )

        # the second stage of initialization
        self.initialization2nd(x, Y, par)


    def initialization2nd(self, x, Y, par):
        ''' second stage of initialization
        The initialization is split into two part for the scenario where 
        the objective function can not be passed as a Python function.
        '''
        # centering origin
        self.center = x
        self.fcenter = self.f[-1]
        # self.centeridx = self.m

        # centered sample
        self.Ycentered = Y

        # centered function value
        self.fcentered = self.f[:-1] - self.fcenter

        # distance of each point in the sample to the current point
        self.distance = np.append(np.linalg.norm(self.Ycentered, axis=1), 0.0)

    
    def update(self, x, f, x1, f1, par, iteration):
        """ The sample is updated every iteration with this function."""

        # add new point
        self.Y = np.concatenate((self.Y, x1.reshape(1, par.n) ))
        self.f = np.append(self.f, f1)

        # update distance
        self.distance = np.linalg.norm(self.Y - x.reshape(1, par.n), axis=1)

        
        # We first delete points. After the deletion, the sample size should
        # be no less tham (par.sample_min - 1), and no more than 
        # (par.sample_max - 1). The "-1" is due to the point we are going 
        # to add into the sample later.

        # indeces of the farthest ones; 
        # We should not delete too many points so that sample size goes under par.sample_min. 
        farthest = np.argsort(self.distance)[-(self.m - par.sample_min):]
        farthest = np.array(farthest)
        # indeces of the ones that are too far
        toofar = np.where(self.distance > par.sample_toremove * par.tr_delta)[0]
        # indeces of the points to be removed
        if toofar.shape[0] == 0 and self.m > par.sample_max:
            # We need to keep the sample size under its maximum. 
            toremove = farthest[-(self.m - par.sample_max):]
        else:
            toremove = np.intersect1d(farthest, toofar)

        # remove points
        self.Y = np.delete(self.Y, toremove, axis=0)
        self.f = np.delete(self.f, toremove)

        # update sample size
        self.m = self.Y.shape[0]
        self.mgood = self.m + toremove.shape[0] - toofar.shape[0]

        # update centering origin 
        self.center = x
        self.fcenter = f
        self.centeridx = np.where((self.Y == self.center).all(axis=1))[0]

        # update centered sample
        self.Ycentered = self.Y[np.arange(self.m) != self.centeridx] - x.reshape(1, par.n)

        # update centered function value
        self.fcentered = self.f[np.arange(self.m) != int(self.centeridx)] - f1

    def addpoints(self, points, fvalues, par):
        # number of new points
        p = points.shape[0]

        # sample size 
        self.m += p

        # sample
        self.Y = np.concatenate((self.Y, points))

        # centered points 
        centeredPoints = points - self.center
        self.Ycentered = np.concatenate((self.Ycentered, centeredPoints))

        # function values
        self.f = np.concatenate((self.f, fvalues))

        # centered function values
        self.fcentered = np.concatenate((self.fcentered, fvalues - self.fcenter))

        # distance to the center
        d = np.linalg.norm(centeredPoints, axis=1)
        self.distance = np.concatenate((self.distance, d))

        # number of sample points within the acceptable range
        self.mgood += sum(d < par.tr_delta * par.sample_toremove)



def quadratic_terms(x):
    n = len(x)
    x = x[np.newaxis] # turn x from 1D to a 2D array
    temp = x * x.T - (1 - 1/np.sqrt(2)) * np.diag(pow(x, 2)[0]) 
    
    return temp[np.tril_indices(n)]